types/validator_set.go

```diff
@@ -366,6 +385,202 @@ func (vals *ValidatorSet) Iterate(fn func(index int, val *Validator) bool) {
+		}
+
+		overflow := updatedVotingPower > MaxTotalVotingPower || updatedVotingPower < 0
+		if overflow {
+			err := fmt.Errorf(
+				"failed to add/update validator %v. total voting power would exceed the max allowed %v",
+				valUpdate,
+				MaxTotalVotingPower)
+			return err
+		}
+	}
+	// Loop again and update the proposerPriority for newly added and updated validators
+	for _, valUpdate := range updates {
+		address := valUpdate.Address
+		_, val := vals.GetByAddress(address)
+		if val == nil {
+			// add val
+			// Set ProposerPriority to -C*totalVotingPower (with C ~= 1.125) to make sure validators can't
+			// un-bond and then re-bond to reset their (potentially previously negative) ProposerPriority to zero.
+			//
+			// Contract: updatedVotingPower < MaxTotalVotingPower to ensure ProposerPriority does
+			// not exceed the bounds of int64.
+			//
+			// Compute ProposerPriority = -1.125*totalVotingPower == -(updatedVotingPower + (updatedVotingPower >> 3)).
+			valUpdate.ProposerPriority = -(updatedVotingPower + (updatedVotingPower >> 3))
+		} else {
+			valUpdate.ProposerPriority = val.ProposerPriority
+		}
+	}
+
+	return nil
+}
+
+// Merges the vals' validator list with the updates list.
+// When two elements with same address are seen, the one from updates is selected.
+// Expects updates to be a list of updates sorted by address with no duplicates,
+// and to have been validated with validateUpdates().
+// The validator's priorities in 'updates' should be already computed.
+func (vals *ValidatorSet) applyUpdates(updates []*Validator) {
+
+	existing := make([]*Validator, len(vals.Validators))
+	copy(existing, vals.Validators)
+
+	merged := make([]*Validator, len(existing) + len(updates))
+	i := 0
+
+	for len(existing) > 0 && len(updates) > 0 {
+		if bytes.Compare(existing[0].Address, updates[0].Address) < 0 {
+			merged[i] = existing[0]
+			existing = existing[1:]
+		} else {
+			merged[i] = updates[0]
+			if bytes.Equal(existing[0].Address, updates[0].Address) {
+				// validator present in both, so advance existing also
+				existing = existing[1:]
+			}
+			updates = updates[1:]
+		}
+		i++
+	}
+	for j := 0; j < len(existing); j++ {
+		merged[i] = existing[j]
+		i++
+	}
+	for j := 0; j < len(updates); j++ {
+		merged[i] = updates[j]
+		i++
+	}
+
+	vals.Validators = merged[:i]
+	vals.totalVotingPower = 0
+}
+
+// Checks that the validators to be removed are part of the validator set.
+func verifyRemovals(deletes []*Validator, vals *ValidatorSet) error {
+
+	for _, valUpdate := range deletes {
+		address := valUpdate.Address
+		_, val := vals.GetByAddress(address)
+		if val == nil {
+			return fmt.Errorf("failed to find validator %X to remove", address)
+		}
+	}
+	return nil
+}
+
+// Removes the validators specified in 'deletes' from validator set 'vals'.
+// Should not fail as verification has been done before.
+func (vals *ValidatorSet)applyRemovals(deletes []*Validator) {
```

Comment:
Formatting seems off:
```suggestion
func (vals *ValidatorSet) applyRemovals(deletes []*Validator) {
```

Reply:
You could add goimports or gofmt to your workflow/IDE to have the code formatted in an idiomatic way.

Reply:
Ouch...I just switched to trying GoLand (from VSCode) and forgot to add gofmt to the file watchers...thanks!